src/libravi/RaviSignalingConnection.ts

```diff
@@ -396,24 +396,45 @@ class RaviSignalingWebSocketImplementation {
     // stateChangeHandlers.
     // (We can't set these until we attempt to open the
     // WebSocket, because there's no other WebSocket constructor.)
-    this._webSocket.addEventListener('open', function(event: any) { signalingConnection._handleStateChange(event, RaviSignalingStates.OPEN); });
-    this._webSocket.addEventListener('error', function(event: any) { signalingConnection._handleStateChange(event, RaviSignalingStates.ERROR); });
-    this._webSocket.addEventListener('close', function(event: any) { 
-        if (event && event.code && event.code > 4000) {
-            // This "close" event is really an error, because we're
-            // returning one of our custom error codes. Treat it as such.
-            RaviUtils.err("_handleStateChange: signaling error code " + event.code + ":  " + event.reason, "RaviSignalingConnection");
-            signalingConnection._handleStateChange(event, RaviSignalingStates.ERROR);
-        } else {
-            signalingConnection._handleStateChange(event, RaviSignalingStates.CLOSED);
-        }
-    });
+    this._webSocket.addEventListener('open', this._onWebSocketOpen);
+    this._webSocket.addEventListener('error', this._onWebSocketError);
+    this._webSocket.addEventListener('close', this._onWebSocketClose);
 
     // Any additional messaging handling gets done by the main
     // RaviSignalingConnection's messageHandlers
     this._webSocket.addEventListener('message', function(event: any) { signalingConnection._handleMessage(event); });
   }
   
+  /**
+   * @private
+   */
+  _onWebSocketOpen = (function(event: any) {
+    this._raviSignalingConnection._handleStateChange(event, RaviSignalingStates.OPEN);
+  }).bind(this);
+
+  /**
+   * These handlers need to be named references so that we can easily remove them from the
+   * websocket when we're closing it.
+   * @private
+   */
+  _onWebSocketError = (function(event: any) {
+    this._raviSignalingConnection._handleStateChange(event, RaviSignalingStates.ERROR);
+  }).bind(this);
+
+  /**
+   * @private
+   */
+  _onWebSocketClose = (function(event: any) {
+    if (event && event.code && event.code > 4000) {
+      // This "close" event is really an error, because we're
+      // returning one of our custom error codes. Treat it as such.
+      RaviUtils.err("_handleStateChange: signaling error code " + event.code + ":  " + event.reason, "RaviSignalingConnection");
+      this._raviSignalingConnection._handleStateChange(event, RaviSignalingStates.ERROR);
+    } else {
+      this._raviSignalingConnection._handleStateChange(event, RaviSignalingStates.CLOSED);
+    }
+  }).bind(this);
```

Comment:
A matter of style: This code is completely legit, but there are a few (debatable) reasons why we might want to just have the constructor set this.onWebSocketClose = function(...etc) {...};, without the bind.:
- Most Javascript programmers don't know exactly what bind does, nor the difference between function and fat arrow.  Elsewhere we have made the specific decision to not used advanced features that programmers might not know, even when it would make the code clearer. (In this case, the code is in a private method, so arguably that's ok.)
- Risk of platform problems. As far as I know, all browsers that we target support _instance_ variables initialized with this syntax (regardless of whether they are functions), but I happen to know that Safari doesn't support static/_class_ variables to be initialized this way. So I'm a little nervous about this one.

Reply:
Update: CURRENT safari supports this syntax for both instance AND class vars, which makes me wonder (and unable to test) whether older (but recent) safari had indeed supported this syntax for instance variables!

Reply:
I don't feel particularly strongly one way or the other, I just personally like `bind`s. :-) Happy to change if you think it's necessary!